src/Model_ELBDM/CPU_ELBDM/CPU_ELBDMSolver_HJ.cpp

```diff
@@ -0,0 +1,588 @@
+
+#     ifdef CONSERVE_MASS
+      CGPU_SHARED real  s_Flux               [CGPU_FLU_BLOCK_SIZE_Y]                            [HYB_NXT];         // the average density fluxes
+#     endif // #  ifdef CONSERVE_MASS ... # else
+
+#     ifdef __CUDACC__
+//    use two-dimensional thread blocks in GPU mode
+      const uint tx            = threadIdx.x;
+      const uint ty            = threadIdx.y;
+      const uint tid           = ty * CGPU_FLU_BLOCK_SIZE_X + tx;                // thread ID within block
+      const int  bx            = blockIdx.x;
+#     else
+//    every block just has a single thread with temporary memory on the stack in CPU mode
+      const uint tx            = 0;
+      const uint ty            = 0;
+      const uint tid           = 0;
+//    in CPU mode, every thread works on one patch group at a time and corresponds to one block in the grid of the GPU solver
+#     pragma omp for schedule( runtime )
+      for (int bx=0; bx<NPatchGroup; bx++)
+#     endif
+      {
+
+         uint Idx, Idx1, Idx2;            // temporary indices used for indexing column updates, writing data to g_Fluid_In, g_Fluid_Out
+#        ifdef CONSERVE_MASS
+         uint Idx3;                       // temporary index used for writing data to g_Flux
+#        endif // # ifdef CONSERVE_MASS
+         uint si, sj;                     // array indices used in the shared memory array
+         uint NStep;                      // number of iterations for updating each column
+         uint j,k;                        // (j,k): array indices used in g_Fluid_In
+         uint Column0                   = 0;                                              // the total number of columns that have been updated
+         uint NColumnOnce               = MIN( NColumnTotal, CGPU_FLU_BLOCK_SIZE_Y );     // number of columns updated per iteration
+         const bool IsCompletelyRefined = g_IsCompletelyRefined[bx];
+
+
+//       loop over all data columns
+         while ( Column0 < NColumnTotal )
+         {
+//          1. load data into shared memory
+            CELL_LOOP(HYB_NXT, 0, 0) {
+               j = j_gap + ( sj + Column0 ) % size_j;
+               k = k_gap + ( sj + Column0 ) / size_j;
+
+//             1.1 determine the array indices for loading global memory data along different directions
+               Idx1 = get1D1( k, j, si, XYZ );
+
+//             1.2 load the interior data into shared memory at time_level 0
+               s_In[sj][0][DENS][si] = g_Fluid_In[bx][DENS][Idx1];
+               s_In[sj][0][PHAS][si] = g_Fluid_In[bx][PHAS][Idx1];
+
+#              ifdef CONSERVE_MASS
+               s_Flux[sj][si] = 0;
+#              endif
+
+               s_HasWaveCounterpart[sj][si]  = g_HasWaveCounterpart[bx][Idx1];
+            }
+
+//          1.4 sync data read into S_In
+#           ifdef __CUDACC__
+            __syncthreads();
+#           endif // # ifdef __CUDACC_
+
+//          first order update
+            if ( IsCompletelyRefined )
```

Comment:
Comment on the reason for adopting a first-order scheme when `IsCompletelyRefined == true`.